### PR TITLE
[httpbp]base64 encoding decoding edgecontext

### PR DESCRIPTION
**Pull request by suzie-su**
## Description
Since our current edgecontext header transfer is using base64encoded so we need to keep it consistent to attach to the header with base64 format
👓 @pacejackson @fishy

### CODE REVIEW

File: httpbp/headers.go [hunk 1]
```diff
@@ -67,11 +69,21 @@ type EdgeContextHeaders struct {
...
+	}
 	return EdgeContextHeaders{
-		EdgeRequest: h.Get(EdgeContextHeader),
+		EdgeRequest: string(ec),
 	}
 }
 
+// AttachEdgeContextHeaders attach EdgeRequestContext into request/response header
+func AttachEdgeContextHeaders(ec *edgecontext.EdgeRequestContext, h http.Header) {
+	ecencoded := base64.StdEncoding.EncodeToString([]byte(ec.Header()))
```
Comment (pacejackson) [inline httpbp/headers.go:83]: typo:

```suggestion
	encoded := base64.StdEncoding.EncodeToString([]byte(ec.Header()))
```

File: httpbp/headers.go [hunk 2]
```diff
@@ -67,11 +69,21 @@ type EdgeContextHeaders struct {
...
+		ec = []byte("")
+	}
 	return EdgeContextHeaders{
-		EdgeRequest: h.Get(EdgeContextHeader),
+		EdgeRequest: string(ec),
 	}
 }
 
+// AttachEdgeContextHeaders attach EdgeRequestContext into request/response header
+func AttachEdgeContextHeaders(ec *edgecontext.EdgeRequestContext, h http.Header) {
```
Comment (pacejackson) [inline httpbp/headers.go:82]: Maybe call this `SetEdgeContextHeader`?

File: httpbp/headers.go [hunk 3]
```diff
@@ -67,11 +69,21 @@ type EdgeContextHeaders struct {
 // NewEdgeContextHeaders returns a new EdgeContextHeaders object from the given
 // HTTP headers.
 func NewEdgeContextHeaders(h http.Header) EdgeContextHeaders {
+	ec, err := base64.StdEncoding.DecodeString(h.Get(EdgeContextHeader))
+	if err != nil {
```
Comment (fishy) [inline httpbp/headers.go:73]: I'm not so sure that we want to throw the error away.

`NewEdgeContextHeaders` is called by `TrustEdgeContext`, and if the decoding failed, `TrustEdgeContext` should also report failure.

Although in most cases returning an empty edge request here will also cause `TrustEdgeContext` to fail, I still prefer this to be done explicitly.

**Comment by suzie-su** [comment]
change notes:
1. renaming
2. explicit the error and and return trust edge context false when error happens
3. use w http.ResponseWriter as argument instead of http.header for SetEdgeContextHeader

File: httpbp/headers.go [hunk 4]
```diff
@@ -66,10 +68,19 @@ type EdgeContextHeaders struct {
...
-		EdgeRequest: h.Get(EdgeContextHeader),
-	}
+//
+// The edge context header using base64 encoding for http transport, so it needs decode first
+func NewEdgeContextHeaders(h http.Header) (EdgeContextHeaders, error) {
+	ec, err := base64.StdEncoding.DecodeString(h.Get(EdgeContextHeader))
+	return EdgeContextHeaders{EdgeRequest: string(ec)}, err
+}
+
+// SetEdgeContextHeader attach EdgeRequestContext into request/response header
```
Comment (pacejackson) [inline httpbp/headers.go:78]: I think this is just `response header` now.

File: httpbp/headers_test.go [hunk 5]
```diff
@@ -278,7 +289,29 @@ func TestTrustHeaderSignatureSignAndVerify(t *testing.T) {
 		},
 	)
 }
+func TestInvalidEdgeContextHeader(t *testing.T) {
```
Comment (fishy) [inline httpbp/headers_test.go:292]: nit: Add a blank line between functions.

File: httpbp/headers_test.go [hunk 6]
```diff
@@ -278,7 +289,29 @@ func TestTrustHeaderSignatureSignAndVerify(t *testing.T) {
...
+	ecSignature, err := truster.SignEdgeContextHeader(
+		httpbp.EdgeContextHeaders{EdgeRequest: "=="},
+		time.Minute,
+	)
+	invalidEdgeContextHeader.Header.Set(httpbp.EdgeContextSignatureHeader, ecSignature)
+
+	if truster.TrustEdgeContext(invalidEdgeContextHeader) {
+		t.Error("trust mismatch, expected False, got True")
+	}
+}
```
Comment (fishy) [inline httpbp/headers_test.go:314]: Here, too.

**Pull request merged**